### PR TITLE
Upgrade to Meteor 2.6

### PR DESCRIPTION
**Pull request by imajus**
This PR is based on the outdated and never completed [upgrade-meteor-2.3](https://github.com/wekan/wekan/tree/upgrade-meteor-2.3) branch. I've sorted out all the issues I've faced during the upgrade and tested the results locally for a while. 

A few points to consider:

1. I've considered the original intent to use `ostrio:i18n` as a mistaken approach so instead I've introduced a custom TAPi18n drop-in replacement.
1. Covered that custom TAPi18n implementation with automated tests.
1. There could be some traces of outdated commits left which may lead to some unused Meteor / NPM dependenices. It was pretty hard to select which ones have to stay during rebasing process.
1. Fixed [that](https://github.com/wekan/wekan/issues/3881#issuecomment-1046059803) `BlazeComponent` issue which had stalled the previous upgrading attempt.
1. I could try to rebase on the tip of the `master` branch to get clearer diff for review. Let me know if there's a need in that.

### CODE REVIEW

**Comment by ocdtrekkie** [comment]
I am worried about meteor/meteor#11666 cropping up here.

File: client/components/users/userAvatar.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import Avatars from '/models/avatars';
 import Users from '/models/users';
 import Org from '/models/org';
 import Team from '/models/team';
+import { formatFleURL } from 'meteor/ostrio:files/lib';
```
Comment (xet7) [inline client/components/users/userAvatar.js:6]: Is that formatFleURL correct? Or is there typo, should it be formatFileURL?

File: models/attachments.js [hunk 2]
```diff
@@ -1,267 +1,88 @@
...
-      // this preserves the name so that notifications can be meaningful after
-      // this file is removed
-      attachmentName: doc.original.name,
-      boardId: doc.boardId,
-      cardId: doc.cardId,
-      listId: doc.listId,
-      swimlaneId: doc.swimlaneId,
-    });
+  Meteor.startup(() => {
+    Attachments.collection._ensureIndex({ cardId: 1 });
```
Comment (xet7) [inline models/attachments.js:85]: Should this be createIndex?

File: models/attachments_old.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  // We should use `beforeWrite`.
+  ...defaultStoreOptions,
+});
+AttachmentsOld = new FS.Collection('attachments', {
+  stores: [store],
+});
+
+if (Meteor.isServer) {
+  Meteor.startup(() => {
+    AttachmentsOld.files._ensureIndex({ cardId: 1 });
```
Comment (xet7) [inline models/attachments_old.js:33]: Should this be createIndex?

File: package-lock.json [hunk 4]
```diff
@@ -344,6 +424,11 @@
         "mermaid": "^8.13.8"
       }
     },
+    "abbrev": {
```
Comment (xet7) [inline package-lock.json:427]: It seems that this small package abbrev was published 5 years ago https://www.npmjs.com/package/abbrev . Does it still work? Or does it need this https://github.com/npm/abbrev-js/pull/34 ? What is the chance of getting new version of 5 year old package?

File: sandstorm.js [hunk 5]
```diff
@@ -1,7 +1,9 @@
+import { Meteor } from 'meteor/meteor';
+import { Picker } from 'meteor/communitypackages:picker';
+
 // Sandstorm context is detected using the METEOR_SETTINGS environment variable
 // in the package definition.
-const isSandstorm =
-  Meteor.settings && Meteor.settings.public && Meteor.settings.public.sandstorm;
+const isSandstorm = Meteor.settings?.public?.sandstorm;
```
Comment (xet7) [inline sandstorm.js:6]: What is this new syntax?

Comment (mfilser) [reply sandstorm.js:6]: It's optional chaining

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: .meteor/packages [hunk 6]
```diff
@@ -17,23 +17,15 @@ es5-shim@4.8.0
...
 dburles:collection-helpers
 idmontie:migrations
-matb33:collection-hooks
 matteodem:easy-search
-mongo@1.11.0
+mongo@1.14.0
 mquandalle:collection-mutations
 
 # Account system
-kenton:accounts-sandstorm
```
Comment (xet7) [inline .meteor/packages:30]: But but... removing Sandstorm accounts support???

File: .meteor/packages [hunk 7]
```diff
@@ -17,23 +17,15 @@ es5-shim@4.8.0
...
-service-configuration@1.0.11
-useraccounts:unstyled
-useraccounts:flow-routing
-wekan-ldap
-wekan-accounts-cas
-wekan-accounts-oidc
```
Comment (xet7) [inline .meteor/packages:36]: wekan-accounts-oidc has support for many custom OIDC/OAuth2 login types, including:
- https://github.com/wekan/wekan/wiki/Oracle-OIM
- https://github.com/wekan/wekan/wiki/ADFS
- https://github.com/wekan/wekan/wiki/OAuth2#auth

Is there some real working replacement available somewhere?

File: .meteor/packages [hunk 8]
```diff
@@ -17,23 +17,15 @@ es5-shim@4.8.0
...
+mongo@1.14.0
 mquandalle:collection-mutations
 
 # Account system
-kenton:accounts-sandstorm
-service-configuration@1.0.11
-useraccounts:unstyled
-useraccounts:flow-routing
-wekan-ldap
-wekan-accounts-cas
```
Comment (xet7) [inline .meteor/packages:35]: Is there any CAS replacement available?

File: .meteor/packages [hunk 9]
```diff
@@ -17,23 +17,15 @@ es5-shim@4.8.0
...
-mongo@1.11.0
+mongo@1.14.0
 mquandalle:collection-mutations
 
 # Account system
-kenton:accounts-sandstorm
-service-configuration@1.0.11
-useraccounts:unstyled
-useraccounts:flow-routing
-wekan-ldap
```
Comment (xet7) [inline .meteor/packages:34]: Is there any LDAP replacement available?

**Comment by xet7** [comment]
From @ocdtrekkie 

> I am worried about [meteor/meteor#11666](https://github.com/meteor/meteor/issues/11666) cropping up here.

From @xet7

Yes, on Sandstorm there is no MongoDB 5 available. With current version of Meteor, releasing Sandstorm WeKan version works. Before that was a long time when it was broken. Anyway, I'll try is there some way to keep everything working when trying to upgrade.

**Comment by xet7** [comment]
Anyway, I'm In Progress of fixing merge conflicts and trying does this work, on new branch.

**Comment by xet7** [comment]
I merged all to new branch upgrade-meteor, and updated to newest Meteor 2.7.1 . It seems to work with password login, adding boards, etc, but there could be something that does not work.

@imajus @mfilser @helioguardabaxo Please test.

**Comment by mfilser** [comment]
Great, thanks to everyone.
I will test in this week.
Do i have to update MongoDB to 5.x ?

**Comment by xet7** [comment]
@mfilser 

Running `meteor` runs embedded MongoDB 5.

**Comment by xet7** [comment]
@mfilser 

This shows when running `meteor`:

meteor

**Comment by mfilser** [comment]
@xet7
What does it mean "embedded MongoDB"?

I have a own MongoDB instance running for Wekan.

**Comment by xet7** [comment]
@mfilser 

When running command `meteor`, during development, Meteor runs MongoDB database that is installed with running `npm install meteor`.

When running `start-wekan.sh`, Docker etc, then bundle at `wekan/.build/bundle` uses MongoDB database defined at MONGO_URL.

**Comment by mfilser** [comment]
Ok, thanks, i will try it.

**Comment by xet7** [comment]
@mfilser 

That embedded MongoDB runs at next port. For example, if running `meteor --port 4000` then you can connect to MongoDB with command `mongo --port 4001`. To delete development MongoDB database, it's possible to use commands `rm -rf .meteor/local`.

**Comment by xet7** [comment]
To build WeKan, for example this works:
```
git clone https://github.com/wekan/wekan
cd wekan
git checkout --branch upgrade-meteor
npm install
./rebuild-wekan.sh
```
And option 3.

**Comment by mfilser** [comment]
I would prefer to have a copy of the production database during development

**Comment by xet7** [comment]
@mfilser 

Then you can define MONGO_URL=... after command meteor.

**Comment by xet7** [comment]
When I try to upgrade to beta https://github.com/meteor/meteor/pull/11999 , I get this error:
```
$ meteor update --release 2.7.2-beta.0
=> Errors while initializing project:                                           
                                              
While selecting package versions:
error: Conflict: Constraint mongo-decimal@0.1.3 is not satisfied by
mongo-decimal 0.1.2.
Constraints on package "mongo-decimal":
* mongo-decimal@0.1.2 <- minimongo 1.8.0 <- aldeed:collection2-core 1.2.0 <-
aldeed:collection2 2.10.0
* mongo-decimal@0.1.2 <- minimongo 1.8.0 <- matb33:collection-hooks 1.1.2
* mongo-decimal@0.1.3 <- mongo 1.15.0-beta272.0
```

**Comment by xet7** [comment]
From @ocdtrekkie

I am worried about https://github.com/meteor/meteor/issues/11666 cropping up here.

From @ocdtrekkie at IRC Libera.Chat sandstorm:

Looks like the Meteor upgrade issue might be fixed as of Meteor 2.7.2? There is a PR that mentions it which involves an option that I believe automatically fixes the indexes. https://github.com/meteor/meteor/pull/11995 is the key bit.
Includes documentation for the option that would need to be set.

**Comment by xet7** [comment]
TODO, try to fix these that do not work in Meteor 2.7.1 WeKan:
- [ ] Adding label and label color
- [ ] Languages menu does not show all language names
- [ ] Changing from English to another language does not work

**Comment by mfilser** [comment]
I would also suggest to add code to test if migration to Meteor-Files was already done, if not, display a error message and abort starting Wekan.

If yes, this code should be at:
https://github.com/wekan/wekan/blob/42b05b263d57ec6745a342bcad74aa566e4531d2/server/00checkStartup.js

**Comment by mfilser** [comment]
... because i think cfs:* packages are removed in this PR, or?

**Comment by xet7** [comment]
@mfilser 

In this PR, WeKan forked cfs was removed from wekan/packages/, and old cfs:* added here:

- https://github.com/wekan/wekan/blob/upgrade-meteor/.meteor/packages#L20
- https://github.com/wekan/wekan/blob/upgrade-meteor/.meteor/packages#L70
- https://github.com/wekan/wekan/blob/upgrade-meteor/.meteor/packages#L80

**Comment by mfilser** [comment]
@xet7
Yes, that's what i meant. So if there isnt any code which ensures to upgrade to a previous Wekan version, e. g. 6.18, Migration can't be done anymore with this PR.

File: package-lock.json (hunk 4 above)
Comment (imajus) [reply package-lock.json:427]: Not sure. It's there as a dependency for `@mapbox/node-pre-gyp` package. The latter was introduced by @xet7 in the commit `4e72dadc6` made on Sat Oct 9 17:00:37 2021 +0300.

File: .meteor/packages (hunk 6 above)
Comment (imajus) [reply .meteor/packages:30]: Not sure. This was removed by @xet7 in the commit `40265144` made on Fri Jun 25 14:11:53 2021 +0300.

**Comment by imajus** [comment]
I see that there are some packages which was removed temporary during previous Meteor 2.3 upgrade attempt, but they are necessary for system to work:
- kenton:accounts-sandstorm
- wekan-ldap
- wekan-accounts-cas
- wekan-accounts-oidc

**Comment by imajus** [comment]
> * `meteor` terminal log shows `i18next: languageChanged en`, list of languages etc language output

@xet7 this is not the issue. I've intentionally activated `i18next` debugging in development environment. I think it's generally a rule of thumb to activate all the debugging output in the development environment.

**Comment by imajus** [comment]
> @xet7 Yes, that's what i meant. So if there isnt any code which ensures to upgrade to a previous Wekan version, e. g. 6.18, Migration can't be done anymore with this PR.

Sorry I don't get what you meant by that. All the necessary dependencies for running GridFS related migrations are present in this PR. I don't foresee any problems with upgrading from CollectionFS to Meteor-Files while running this version of the app.

**Comment by imajus** [comment]
> * Languages menu does not show all language names

@xet7 yep, that's what I've witnessed as well. I thought that maybe it was an issue before the PR? For some reason, `useraccounts:unstyled` Meteor package doesn't pick up the translations provided by the `meteor-accounts-t9n` package. 

The general translations used in the application worked fine and updated reactively when l changed a different language for the user in the Boards UI.

>  Languages menu does not show all language names

This is pretty interesting thing. So apparently the application had a lot more translations than was listed in the authentication page. Not sure about the language selection in the Boards UI, maybe we had a limited languages listing there too?

I've made it to list all of the supported languages but had to skips the labels for those which weren't listed there before. Not sure what approach should be taken here. Should we instead list only those which had been listed there previously? Or should we add proper labels to the languages currently lacking of such?

**Comment by imajus** [comment]
I've fixed a few listed issues:

- [x] Languages menu does not show all language names
- [x]  Changing from English to another language does not work

**Comment by xet7** [comment]
@imajus 

I have added these changes to meteor-upgrade branch. For future changes, please add new PRs for meteor-upgrade branch. I'm currently merging some newest changes from master to meteor-upgrade branch.

Moved to #4485

**Pull request closed without merging**